docs/resiliency-checklist.md

```diff
@@ -83,12 +83,24 @@ It is preferable to scale horizontally than vertically. Use smaller Request and
 
 Make use of the [Quality of Service](https://docs.openshift.com/container-platform/3.11/dev_guide/compute_resources.html#quality-of-service-tiers) options available to ensure your pods are correctly resourced and treated appropriately by the scheduler.
 
-If your pod has its Requests and Limits set to `0` it will run BestEffort and simply use whatever spare capacity is available on its node. This can be good for things like batch jobs that want to go as fast as they can, but don't care about being slowed down from time to time. It will also be the first to be evicted from an overloaded node. Assigning `0` as a request or limit must be done through the CLI or directly in the manifest. 
+If your pod has its Requests and Limits set to `0` it will run BestEffort and simply use whatever spare capacity is available on its node. This can be good for things like batch jobs that want to go as fast as they can, but don't care about being slowed down from time to time. It will also be the first to be evicted from an overloaded node. Assigning `0` as a request or limit must be done through the CLI or directly in the manifest.
 
 If your pod has Limits higher than its Requests it will run Burstable and get at minimum the Requested amount and by burst up to its Limit depending on how busy the node is with other pods.
 
 If your pod has the Requests and Limits set to the same value then it will run Guaranteed QoS class. It will have preferential access to compute resources and will be the last to be evicted if a node should become overloaded. This is best for apps in the `prod` namespace as they need the best uptime.
 
+### A Well Behaved App
+
+It is important to ensure your app is well behaved and doesn't impede the cluster's operators work or impact the cluster's ability to heal.
+
+Review the Kubernetes docs on [pod terminations](https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-termination) to understand the full sequence of events. Ensure that your pods will gracefully exit. This means not setting the `terminationGracePeriod` too high, and ensuring your containers command properly listens for the `SIGTERM` and responds.
+
+When draining nodes for patching, the ops team will cap the grace period to 600s and force kill anything left running after that.
```

Comment:
While ten minutes sounds like a really long time–and genuinely is a generous figure to shut down a tcp pool–it's not necessarily enough time for a stateful application like a database.

From the above, it might take nearly five minutes just to hang up the open client tcp connections, and if those connections were holding transactions or locks open on the database we won't be able to checkpoint that state until the connections are released. So for a stateful application like a patroni cluster we don't just need to hup all connections to our leader, we then need to ensure the write-ahead log stream has been successfully transmitted to all followers, particularly the follower we are about to elect as leader. For our follower nodes, we can kill them off willy-nilly just like any other stateless application since their state is always lagging the elected leader. But for leader nodes, if we fail to observe a complete graceful shutdown then we do not finish transmitting our WAL to our leader-elect and we reliably produce a data loss equivalent to the steady state replication delay between the leader and follower.

The the application connected to this backing store will determine if this kind of regular data loss is acceptable or not. If an application author using postgres, for instance, and remembers to `SET LOCAL synchronous_commit = 'remote_write';` on their high-value transactions or specially configures their writer user `ALTER USER transaction_user SET synchronous_commit= ON;` then the can guarantee they don't lose a few million in high value transactions during that window of data loss: the psql client is aware of the follower status and does not commit the transaction until it is synchronously replicated to all connected followers. There is a significant performance hit to this configuration, which is why it is not the default. But there is also an opportunity for a significant gain: if we force synchronous writes our followers are always in sync with our leaders and we can connect our apps to them as read-replicas without any risk of inconsistency. Is there an opportunity here for a dive into best practices for stateful workloads that optimize for both resiliency of our apps and resiliency of the cluster?

What about time-bound processes? Sometimes, particularly for stateful deployment processes that interact with things like the k8s api that do not have any transaction support, it can be useful to guarantee a process completes its execution without eviction. How might we handle that use case? Helm 3, for instance, is officially supported in OCP4 and ends up in an undefined state if it is force-killed partway through execution. Would it ever be acceptable to set our `terminationGracePeriod` even with our `activeDeadlineSeconds`?

Reply:
@jefkel you know more about stateful apps than I do, any input on?
> Is there an opportunity here for a dive into best practices for stateful workloads that optimize for both resiliency of our apps and resiliency of the cluster?

I'll have to do some testing in January on how time-bound pods handle drains. 

> Would it ever be acceptable to set our terminationGracePeriod even with our activeDeadlineSeconds?

Depends on how big your `activeDeadlineSeconds` is I guess? If its a cronjob that needs to run for an hour, then we probably wont want that. But if its a helm task that is only going to run for 5 min then should be fine to set them equal and just ensure it completes.

Reply:
Feels like we may be over-thinking the optimization of stateful workloads.  In non-container flows, a comparable situation is when a physical host loses power (assuming the host has some UPS support).  If your physical host has a 10m expectancy of available battery power, your application's stateful behavior is forced to accommodate this physical constraint.  (regardless of how uncomfortable this is)

In our conversations about containerizing stateful workload, I would suggest that the "Termination Grace Period" be treated in a similar way to "Minutes available of battery power", ie: as a physical constraint.  I believe at this point, the aim is to grant every pod a 10m grace (correlated to a physical limitation).   Increasing that constraint needs to involve the hosting service as this will directly impact costs re: maintenance, but still could be looked on (from an application design perspective) as a physical constraint.

For stateful workloads as well as time-bound processes, ensuring the ability to terminate/restart with minimal data-loss or service impact within (or faster than) that time constraint is a great place to start.  This is not a new constraint, and has many different approaches for applications based on the application architectures.  (I would be happy to dig in for individual cases, but would shy away from a general "best practice")

re: matching activeDeadlineSeconds - That's like saying we need to have enough battery power available for a physical host to guarantee that the longest job never fails.   Not something I've ever seen used in a practical situation.  Generally the approach is from the other direction: Make sure your long running job is designed so that it can restart and/or continue from where it left off in the case of a disruption.